Use DEFLATE compression when rendering AuthnRequest

Philonous opened the pull request:
This PR changes AuthnRequest rendering to compress using DEFLATE before Base64-encoding. 

* Required by SAML HTTP redirect binding if no other encoding is specified
* Uncompressed request is in fact rejected by Keycloak with `Invalid Request` 
* The use of HTTP redirect binding is implied by the the comment:
> which should be used as a query parameter named @SAMLRequest@

This change might break @fumieval's use case, if so we should probably make it configurable, but this is the explicitly mentioned standard behaviour

Reference: 

> A query string parameter named SAMLEncoding is reserved to identify the
encoding mechanism used. If this parameter is omitted, then the value is assumed to be urn:oasis:names:tc:SAML:2.0:bindings:URL-Encoding:DEFLATE

https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf#page=17

Philonous committed:
Add renderUrlEncodingDeflate to Network.Wai.SAML2.Request

* rename renderAuthnRequest to renderBase64
* add render

* Required by default HTTP redirect request

> A query string parameter named SAMLEncoding is reserved to identify the
encoding mechanism used. If this parameter is omitted, then the value is assumed
to be urn:oasis:names:tc:SAML:2.0:bindings:URL-Encoding:DEFLATE

https://docs.oasis-open.org/security/saml/v2.0/saml-bindings-2.0-os.pdf#page=17

Philonous commented:
Should mention: `zlib` is already in the transitive dependencies (at least via [`wai-extra`](https://hackage.haskell.org/package/wai-extra) -> [`streaming-commons`](https://hackage.haskell.org/package/streaming-commons)), so adding it directly doesn't increase the total footprint

fumieval commented:
> The use of HTTP redirect binding is implied by the the comment:

I realised that the module comment is not what I intended; the actual behaviour is for the HTTP POST Binding.

This change breaks HTTP POST Binding (confirmed on both AzureAD and Google Workspace) so I think it should be ~~configurable~~ a separate function

Philonous commented:
I have added a new function `renderUrlEncodingDeflate` which is suitable for HTTP redirect binding and renamed `renderAuthnRequest` (which has become ambiguous) to `renderBase64`. I've also factored out the basic XML rendering as new ~`render`~ `renderXML` function and exported it (makes it easier to re-use for potentially new schemes) 

I've also updated documentation to clarify the use cases. 

Feel free to choose better names, I've been agonizing over them and decided that keeping them short makes them easiest to use when the module is imported qualified. 

Edit: Changed `render` to `renderXML`

fumieval commented:
Thanks, I have no objection to the names. Perhaps `renderXML` is more consistent? (`render` is unlikely to be used in regular scenarios anyway)

Philonous commented:
I renamed `render` to `renderXML` as per your suggestion

Philonous commented:
I seem to have forgotten the `URL-Encoding` part of `URL-Encoding:DEFLATE`. Oops. :sweat_smile:  Pushed a fix.

mbg reviewed (approved):
LGTM. Thanks for the well documented work and thoughtful discussion, @Philonous! Thanks @fumieval for your input here as well! :)
